packages/lib-classifier/src/store/SubjectStore.js

```diff
@@ -5,9 +5,26 @@ import { getBearerToken } from './utils'
 import { filterByLabel, filters } from '../components/Classifier/components/MetaTools/components/Metadata/components/MetadataModal'
 import ResourceStore from './ResourceStore'
 import Subject from './Subject'
+import SingleImageSubject from './SingleImageSubject'
+import SubjectGroup from './SubjectGroup'
 
 const MINIMUM_QUEUE_SIZE = 3
 
+/*
+  see https://github.com/mobxjs/mobx-state-tree/issues/514
+  for advice about using references with types.union.
+*/
+const SingleSubject = types.union(SingleImageSubject, Subject)
+function subjectDispatcher (snapshot) {
+  if (snapshot.subjects) {
+    return SubjectGroup
+  }
+  return SingleSubject
+}
```

Comment:
[notes to self]

I've been trying to wrap my head around this (unions and dispatchers aren't my forte, alas), so I'm going to write out what I can discern from this code:

- the "dispatcher" is just a function whose sole job is to determine what _type of_ Subject is being seen in a particular instance of a union-ed data model.
- Usually the types.union([list_of_types]) data model would be smart enough (?) to figure out the _type_ of the instance from the shape of its data, but the dispatcher is there to make the selection explicit. By default it selects the first matching type from the 'list_of_types'.
- So basically what we have here is...
```
SubjectStore.active('s data model) = (snapshot) => {
  if snapshot has .subjects, return SubjectGroup
  else,
    if snapshot has only 1 image (as determined by .refinement), return SingleImageSubject
    else, return generic Subject 
}
```

If that's the case, then this dispatcher will become an important logical "selection crossroads" for future development, when potential new Subject types are added.

It might be worth noting/broadcasting this info, because otherwise future devs might start encoding the "selection logic" in individual data model definitions, so you might need to dig through 5 layers of specific sub-types+unions+dispatchers to determine why Subject.active is currently of the type SingleImageWithAnimalsInAfricaWithBlueSkiesButNoAardvarks. (Great for compartmentalisation the code, headachey for instantly grokking what's happening.)

Similar notes can be applied for `Subject`'s `get viewer ()` which has a "selection crossroad" and `SubjectGroup`'s `get viewer ()` which overrides it.

Reply:
Yeah, a dispatcher is a function that takes a snapshot and returns a type for that snapshot. It’s used when a type can’t be determined from a snapshot ie. snapshots aren’t unique to models.